Refactor check niimg

AlexandreAbraham opened the pull request:
This PR restore the original behavior of check_niimg and check_niimgs:
* check_niimg loads 3D niimages
* check_niimgs loads 4D niimages

This is a first step toward cleaner file loading in nilearn.

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -92,6 +88,10 @@ def _to_4d(data):
+def _iter_concat_niimgs():
+    pass
+
+
```
:question:

AlexandreAbraham replied:
:swimmer: WIP

AlexandreAbraham replied:
i postponed this change.

GaelVaroquaux replied:
> i postponed this change.

Then you should remove the corresponding 'stub' function.

lesteve commented:
Plenty of failing tests in case you missed it somehow.

AlexandreAbraham commented:
It's WIP at the moment. It's far from being usable.

AlexandreAbraham commented:
So, I fixed the failing tests during lunch.

About the changes:
* 5D images or 4D masks are NOT accepted anymore (they were valid in a few functions)
* loading of 3D or 4D niimg is now explicit. A generic `check_niimg` exists (for 3D or 4D) is however present for explicit polymorphic functions (eg smooth_img)
* A lot of tests were totally wrong in their logic. We have to be more careful about what we test

I suggest that we merge the PR as-is.

Future work:
* `concat_niimgs` is a weird thing that accepts mix of 3D and 4D images. It shouldn't do that.
* Functions to load / iterate on list of niimg would be nice. For example `check_niimgs_3d` and `check_niimgs_4d` with optional parameters to check affines and stuff.
* This PR is only about clarifying difference between 3D and 4D as it was at the beginning (with niimg and niimgs). It does not fix #463 but it should fix #501.

I wanted to throw out this code that was in a branch for a very long time but i have no time to work on the other issues.

AlexandreAbraham commented on nilearn/_utils/niimg.py:
```diff
@@ -29,7 +29,7 @@ def _safe_get_data(img):
 
 
-def load_img(img, dtype=None):
+def load_niimg(img, dtype=None):
```
Renamed it because the input is a niimg but the output is an img.

lesteve replied:
but then I guess the argument should be called `niimg` too ...

GaelVaroquaux replied:
> Renamed it because the input is a niimg but the output is an img.

I am not sure that this is consistent with our naming conventions, and
'smooth_img'.

AlexandreAbraham commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -4,29 +4,36 @@
 
-def check_niimg(niimg, ensure_3d=False):
-    """Check that niimg is a proper niimg. Turn filenames into objects.
+def check_niimg(niimg, atleast_4d=False):
```
This function is here to be explicitely used in polymorphic functions.

banilo replied:
I am not sure that "atleast_4d" is intuitively understood by many users.

GaelVaroquaux replied:
> I am not sure that "atleast_4d" is intuitively understood by many users.

But that's internal. And it corresponds to a numpy convention.

AlexandreAbraham commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -193,22 +223,45 @@ def concat_niimgs(niimgs, dtype=np.float32, accept_4d=False,
 
 
-def check_niimgs(niimgs, accept_3d=False, return_iterator=False):
+def _iter_check_niimg_4d(niimgs):
```
If somebody asks for an iterator on 4d, this iterator will check for everything during the process. Before that, the code could return niimages with different fov. I like this logic and I think that this could be propagated in other parts of the code.

AlexandreAbraham commented on nilearn/input_data/tests/test_base_masker.py:
```diff
@@ -55,5 +55,8 @@ def test_filter_and_mask():
+
     assert_raises_regex(TypeError, "A 3D image is expected", filter_and_mask,
-                        data, mask, {})
+                         data_img, mask_img, {})
```
For example, this test was passing but it was bogus because it called `filter_and_mask` with regular numpy arrays instead of niimg. This shouldn't even work.

GaelVaroquaux replied:
> For example, this test was passing but it was bogus because it called
> filter_and_mask with regular numpy arrays instead of niimg. This
> shouldn't even work.

Agreed

AlexandreAbraham commented on nilearn/tests/test_masking.py:
```diff
@@ -200,26 +200,6 @@ def test_unmask():
             assert_true(t[0].flags["F_CONTIGUOUS"])
             assert_array_equal(t[0], unmasked3D)
 
-    # 5D test
```
5D data is not accepted. I think that if we want to accept it, we should do all the proper verifications that comes with it.

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -134,8 +161,11 @@ def concat_niimgs(niimgs, dtype=np.float32, accept_4d=False,
     """
 
+    if len(niimgs) == 0:
+        raise TypeError('Cannot concatenate empty objects')
```
What if niimgs is an iterator (which seems to be a possibility if you trust the docstring)?

IMHO the current code in master is badly broken with iterators and I would suggest that we only support lists until we have proper tests for iterators.

AlexandreAbraham replied:
This may be a regression so I fixed (it was easy). As said in my comment, I don't make the code iterator friendly. This is a lot of work and I have no time to do it. Plus I prefer to segment things in smaller PRs.

GaelVaroquaux replied:
> IMHO the current code in master is badly broken with iterators and I would
> suggest that we only support lists until we have proper tests for iterators.

Agreed.

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -193,22 +223,45 @@ def concat_niimgs(niimgs, dtype=np.float32, accept_4d=False,
+                       niimg.shape))
+            yield niimg
+        except TypeError as e:
+            raise TypeError("Error while loading image #%d: \n%s"
```
This kind of thing will lose information about the provenance of the error which is very annoying if you are dropping into a debugger. Look at `git grep exc.args +=` for example in our code of how to decorate an existing exception.

AlexandreAbraham replied:
Fixed.

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -237,66 +286,45 @@ def check_niimgs(niimgs, accept_3d=False, return_iterator=False):
-            raise TypeError('An empty object - %r - was passed instead of an '
-                            'image or a list of images' % niimgs)
-        first_img = next(iter(first_img))
-        depth += 1
```
I am so glad to see this slightly baroque code go away ;-)

AlexandreAbraham replied:
:rage2: I will be back

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -4,29 +4,36 @@
+
+def _check_fov(img, affine, shape):
+    """ Return True if img1 and have the given field of view
+        (shape and affine), False elsewhere.
```
Slightly weird sentence, maybe:

```python
""" Return whether `img` has the same field of view as the one given by `shape` and `affine`
"""
```

At the very least img1 -> img.

AlexandreAbraham replied:
Fixed

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -135,7 +162,10 @@ def concat_niimgs(niimgs, dtype=np.float32, accept_4d=False,
+    try:
+        first_niimg = check_niimg(next(iter(niimgs)), atleast_4d=True)
+    except StopIteration:
+        raise TypeError('Cannot concatenate empty objects')
```
maybe ValueError rather than TypeError, since an empty list still has the right type ?

AlexandreAbraham replied:
I kept the original behavior (this was checked higher in the stack before) but I agree with you. Maybe we can change that in another PR as I want this one to have the lowest impact possible on the logic of the code.

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -237,66 +286,45 @@ def check_niimgs(niimgs, accept_3d=False, return_iterator=False):
-        return niimg
+    # Two types of input:
+    # * 4D nifti image
+    # * list of 3D niimgs
```
This is not really consistent with the docstring that says that niimgs can be an iterable ...

AlexandreAbraham replied:
I didn't put it here but iterator should work if you do `return_iterator=True` (not the other case as `concat_niimgs` is not iterator friendly). But, again, the goal of this PR is not to change the behavior of the code when using iterators. It is exactly as broken as before ;)

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -237,66 +286,45 @@ def check_niimgs(niimgs, accept_3d=False, return_iterator=False):
+        niimgs = load_niimg(niimgs)
+        shape = niimgs.shape
+        if len(shape) == 3:
+            raise TypeError("A 4D image is expected, but an image "
```
Should we use 4D Niimg-like object as previously?

AlexandreAbraham replied:
Fixed.

GaelVaroquaux commented:
>   • 5D images or 4D masks are NOT accepted anymore (they were valid in a few
>     functions)

That may be a problem: these things do lie around.

>   • concat_niimgs is a weird thing that accepts mix of 3D and 4D images. It
>     shouldn't do that.

Why?

AlexandreAbraham commented:
> That may be a problem: these things do lie around.

I see no reason to support them in only two functions (reorder_img and unmask). we should support them completely, or not at all.

> Why?

For me, trying to concatenate a 4D image with a 3D image is most likely an error than a wanted behavior.

GaelVaroquaux commented:
>     That may be a problem: these things do lie around.

> I see no reason to support them in only two functions (reorder_img and
> unmask). we should support them completely, or not at all.

Just saying. Users of AFNI will be in big trouble.

>     Why?

> For me, trying to concatenate a 4D image with a 3D image is most likely an
> error than a wanted behavior.

Agreed. So we should rather test for consistency in the number of
dimensions.

AlexandreAbraham commented:
> Just saying. Users of AFNI will be in big trouble.

They already are. As said before, I don't really fix bugs nor add any functionality. I just fixed the fact that check_niimg (now check_niimg_3d) only loads 3D images and check_niimgs (now check_niimg_4d) only loads 4D images. It is true that, in the meantime, I removed the ability of some functions to handle 5d data, but NiftiMasker never did for example.

If we want to support 5D images, it is easy to do that in a lot of functions (smooth_img, reorder_img...). It can be done by modifying check_niimg to accept 5d data. Oh, and it may be possible to handle in MultiNiftiMasker rather easily.

Actually, we should grab some AFNI data to test it.

GaelVaroquaux commented:
>     Just saying. Users of AFNI will be in big trouble.

> They already are.

OK, let's postpone saving them, then.

banilo commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -55,41 +61,62 @@ def check_niimg(niimg, ensure_3d=False):
 
     Its application is idempotent.
     """
+    # If it's an iterator, it's a 4d image
```
Shouldn't we say "iterable" instead? A list also exposes the __iter__ attribute for example.

AlexandreAbraham replied:
Fixed

banilo commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -55,41 +61,62 @@ def check_niimg(niimg, ensure_3d=False):
-    """ Internal function to cast a 3D ndarray to a 4D one by adding a
-        new axis at the end
+def check_niimg_3d(niimg):
+    """Check that niimg is a proper 3D niimg. Turn filenames into objects.
```
Tautological definitions are by definition not a definition.

AlexandreAbraham replied:
A fix has been applied to fix what should be fixed.

banilo commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -55,41 +61,62 @@ def check_niimg(niimg, ensure_3d=False):
+    nilearn because it checks if data have a correct format and loads them if
+    necessary.
+
+    Its application is idempotent.
```
Given the internationality of user pool, perhaps exchange idempotentn by something simpler - "identical"

eickenberg replied:
how is `idempotent` not international??

it doesn't mean "identical". It means that if you apply it n times, where n
>= 1, it is like applying it once.

On Thu, Mar 19, 2015 at 8:58 PM, Danilo Bzdok <notifications@github.com>
wrote:

> In nilearn/_utils/niimg_conversions.py
> <https://github.com/nilearn/nilearn/pull/505#discussion_r26791583>:
>
> > +        and get_affine() methods are present, raise TypeError otherwise.
> > +
> > +    Returns
> > +    -------
> > +    result: 3D Niimg-like object
> > +        Result can be nibabel.Nifti1Image or the input, as-is. It is guaranteed
> > +        that the returned object has get_data() and get_affine() methods.
> > +
> > +    Notes
> > +    -----
> > +    In nilearn, special care has been taken to make image manipulation easy.
> > +    This method is a kind of pre-requisite for any data processing method in
> > +    nilearn because it checks if data have a correct format and loads them if
> > +    necessary.
> > +
> > +    Its application is idempotent.
>
> Given the internationality of user pool, perhaps exchange idempotentn by
> something simpler - "identical"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/505/files#r26791583>.
>

AlexandreAbraham replied:
We could write "Project a niimg in the space of imgs".

banilo commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -221,10 +274,6 @@ def check_niimgs(niimgs, accept_3d=False, return_iterator=False):
-       to 1.
-
     return_iterator: boolean
         If False, a single 4D image is returned. When `niimgs` contains 3D
```
You start out by explaining the default behavior - I would inverse that.

AlexandreAbraham replied:
Fixed.

banilo commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -237,66 +286,45 @@ def check_niimgs(niimgs, accept_3d=False, return_iterator=False):
+        shape = niimgs.shape
+        if len(shape) == 3:
+            raise TypeError("A 4D image is expected, but an image "
+                "with a shape of %s was given." % (shape, ))
```
Why a tuple?

AlexandreAbraham replied:
Not the slightest idea ;). I copy pasted another error message but tuples for single string replacement are all around the code.

banilo commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -237,66 +286,45 @@ def check_niimgs(niimgs, accept_3d=False, return_iterator=False):
+    # * an iterator
+    # * a generator
+
+    # To be iterator/generator friendly, we externalize the verifications in
```
also tautological -> therefore devoid of sense

AlexandreAbraham replied:
why?

banilo commented on nilearn/image/resampling.py:
```diff
@@ -371,9 +371,9 @@ def resample_img(img, target_affine=None, target_shape=None,
-    # noop cases
     img = _utils.check_niimg(img)
 
+    # noop cases
```
You tought me a new thing in coding lingo, thanks.

AlexandreAbraham replied:
You're welcome.

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -237,66 +286,49 @@ def check_niimgs(niimgs, accept_3d=False, return_iterator=False):
+    if isinstance(niimgs, _basestring) or not hasattr(niimgs, "__iter__"):
+        niimgs = load_niimg(niimgs)
+        shape = niimgs.shape
+        if len(shape) == 3:
```
Should you not check `!= 4` rather than `== 3`

AlexandreAbraham replied:
Fixed

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -237,66 +286,49 @@ def check_niimgs(niimgs, accept_3d=False, return_iterator=False):
         else:
-            result = concat_niimgs(niimgs)
-    return result
+            return check_niimg(niimgs)
```
Is this check_niimg needed given that `niimgs` has already been loaded ?

AlexandreAbraham replied:
Fixed

lesteve commented on nilearn/_utils/niimg_conversions.py:
```diff
@@ -193,23 +223,46 @@ def concat_niimgs(niimgs, dtype=np.float32, accept_4d=False,
+            yield niimg
+        except TypeError as exc:
+            exc.args += ('Error encoutered while loading image #%d: \n%s'
+                         % (i, exc.message))
```
You probably forgot a raise after this line to reraise the original exception ...

lesteve replied:
and you don't need exc.message because it is already part of the stack trace.

AlexandreAbraham replied:
Fixed

lesteve replied:
I am a bit wary that this wasn't caught by our tests. Could you add a test where the check_niimg_3d raises a TypeError ? Bonus point if you add one where the field of view is different.

lesteve commented on nilearn/_utils/niimg.py:
```diff
@@ -44,14 +44,14 @@ def load_img(img, dtype=None):
     img: image
```
I guess this should be `niimg` and the type description should be Niimg-like object + URL

AlexandreAbraham replied:
For me, Niimg-like object includes filename.

lesteve replied:
Sorry I misread the diff completely but then a few liines above "or filepath" is not needed, right ?
